optimum/exporters/executorch/integrations.py

```diff
@@ -24,15 +24,274 @@
+    The result of this subgraph should stream directly into the decoder subgraph.
+    """
+
+    def __init__(self, model: torch.nn.Module):
+        super().__init__()
+        self.audio_encoder = model.audio_tower
+        self.mm_projector = model.multi_modal_projector
+        self.intermediate_size = model.config.audio_config.intermediate_size
+        self.audio_token_id = model.config.audio_token_id
+        self.metadata = save_config_to_constant_methods(model.config.text_config, model.generation_config)
+
+    def prepare_export_inputs(self):
+        chunk_length = (
+            self.audio_encoder.config.max_source_positions
+            * self.audio_encoder.conv1.stride[0]
+            * self.audio_encoder.conv2.stride[0]
+        )
+        input_features = torch.rand(3, 128, chunk_length)  # (bsz, features, seq_len)
+
+        max_audio_len = 120  # In s, should be a multiple of 30. TODO: make configurable.
+        max_seq_len = self.metadata.get("get_max_seq_len")
+        dynamic_shapes = {
+            "input_features": {
+                0: torch.export.Dim("enc_batch_size_dim", min=1, max=max_audio_len // 30),
+            },
+        }
+
+        return input_features, dynamic_shapes
+
+    def forward(
+        self,
+        input_features: torch.FloatTensor,
+    ):
+        """
+        Forward pass of the Voxtral encoder module.
+
+        Args:
+            input_features (torch.FloatTensor): Raw audio features with shape (batch_size, features, seq_len).
+
+        Returns:
+            torch.FloatTensor: Combined embeddings with audio tokens replaced by audio embeddings.
+        """
+        audio_outputs = self.audio_encoder(input_features)
+        audio_hidden_states = audio_outputs.last_hidden_state
+        audio_hidden_states = audio_hidden_states.reshape(-1, self.intermediate_size)
+        audio_embeds = self.mm_projector(audio_hidden_states)
+        return audio_embeds
+
+
+class MultiModalTextToTextExportableModule(torch.nn.Module):
+    """
+    A wrapper module for exporting an early fusion multimodal model (image/audio to text) with `torch.export`.
+    This module also ensures that the exported model is compatible with ExecuTorch.
+
+    The module separates the model into three exportable components:
+    1. Token embeddings layer for text encoding
+    2. Multimodal encoder (audio/vision) for processing non-text inputs
+    3. Text decoder for autoregressive generation
+
+    Args:
+        model (torch.nn.Module): The multimodal model to export.
+        modality (str): The input modality type ("audio" or "vision").
+        encoder_name (str): Name of the encoder attribute in the model.
+        use_custom_kv_cache (bool): Whether to use custom key-value caching for optimization.
+        use_custom_sdpa (bool): Whether to use custom scaled dot-product attention.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        modality: str,
```

Comment:
we should definitely look into making it a more generic than only on modality as Mergen said

Reply:
There have been a bunch of comments saying this exact same thing, let's continue the discussion in those above comment threads?